File: node.js/src/slimmed/domain/books/book.ts
import { positiveNumber } from '#core/typing';
import {
  Author,
  AuthorId,
  Chapter,
  CommitteeApproval,
  Format,
  Genre,
  ISBN,
  Reviewer,
  Title,
  Translation,
} from './entities';
import { BookId } from './entities/bookId';
import { IBookFactory } from './factories/bookFactory';
import { InvalidOperationError } from '#core/errors';
import {
  Initial,
  Draft,
  DraftEvent,
  isInitial,
  initialDraft,
  isDraft,
  evolve as evolveDraft,
} from './draft';
import {
  UnderEditing,
  UnderEditingEvent,
  evolve as evolveUnderEditing,
  initial as initialUnderEditing,
  isUnderEditing,
} from './underEditing';
import {
  InPrint,
  InPrintEvent,
  evolve as evolveInPrint,
  initial as initialInPrint,
  isInPrint,
} from './inPrint';
import {
  PublishedBook,
  PublishedEvent,
  evolve as evolvePublished,
  initial as initialPublished,
  isPublished,
} from './published';
import {
  OutOfPrint,
  OutOfPrintEvent,
  evolve as evolveOutOfPrint,
  initial as initialOutOfPrint,
} from './outOfPrint';
import { DomainEvent } from 'src/original/infrastructure/events';

export type Book =
  | Initial
  | Draft
  | UnderEditing
  | PublishedBook
  | InPrint
  | OutOfPrint;

export const evolve = (book: Book, event: BookEvent): Book => {
  switch (event.type) {
    case 'DraftCreated': {
      if (!isInitial(book) || !isDraft(book)) return book;
      return evolveDraft(initialDraft, event);
    }
    case 'ChapterAdded': {
      if (!isDraft(book)) return book;

      return evolve(book, event);
    }
    case 'MovedToEditing': {
      if (!isDraft(book)) return book;

      return evolveUnderEditing(initialUnderEditing, event);
    }
    case 'TranslationAdded':
    case 'TranslationRemoved':
    case 'FormatAdded':
    case 'FormatRemoved':
    case 'ReviewerAdded':
    case 'Approved':
    case 'ISBNSet': {
      if (!isUnderEditing(book)) return book;

      return evolveUnderEditing(book, event);
    }
    case 'MovedToPrinting': {
      if (!isUnderEditing(book)) return book;

      return evolveInPrint(initialInPrint, event);
    }
    case 'Published': {
      if (!isInPrint(book)) return book;

      return evolvePublished(initialPublished, event);
    }
    case 'MovedToOutOfPrint': {
      if (!isPublished(book)) return book;

      return evolveOutOfPrint(initialOutOfPrint, event);
    }

    default: {
      return book;
    }
  }
};

export type BookEvent =
  | DraftEvent
  | UnderEditingEvent
  | InPrintEvent
  | PublishedEvent
  | OutOfPrintEvent;

export type PublishedExternal = DomainEvent<
  'Published',
  {
    bookId: BookId;
    isbn: ISBN;
    title: Title;
    authorId: AuthorId;
  }
>;

export type BookExternalEvent =
  | DraftEvent
  | UnderEditingEvent
  | PublishedExternal
  | OutOfPrintEvent;

export class BookFactory implements IBookFactory {
  create(
    state: State,
    title: Title,
    author: Author,
    genre: Genre | null,
    isbn: ISBN | null,
    committeeApproval: CommitteeApproval | null,
    reviewers: Reviewer[],
    chapters: Chapter[],
    translations: Translation[],
    formats: Format[],
  ): Book {
    switch (state) {
      case State.Writing:
        return {
          status: 'Draft',
          genre,
          chapterTitles: chapters.map((ch) => ch.title),
        };
      case State.Editing: {
        if (genre == null)
          throw InvalidOperationError('Genre should be provided!');

        return {
          status: 'UnderEditing',
          genre,
          isISBNSet: !!isbn,
          isApproved: !!committeeApproval,
          reviewers: reviewers.map((r) => r.id),
          translationLanguages: translations.map((r) => r.language.id),
          formats: formats.map((f) => {
            return { formatType: f.formatType, totalCopies: f.totalCopies };
          }),
        };
      }
      case State.Printing: {
        const totalCopies = formats.reduce(
          (acc, format) => acc + format.totalCopies,
          0,
        );
        return { status: 'InPrint', totalCopies: positiveNumber(totalCopies) };
      }
      case State.Published: {
        const totalCopies = formats.reduce(
          (acc, format) => acc + format.totalCopies,
          0,
        );
        const totalSoldCopies = formats.reduce(
          (acc, format) => acc + format.soldCopies,
          0,
        );

        return {
          status: 'Published',
          totalCopies: positiveNumber(totalCopies),
          totalSoldCopies: positiveNumber(totalSoldCopies),
        };
      }
      case State.OutOfPrint:
        return { status: 'OutOfPrint' };
    }
  }
}

export enum State {
  Writing = 'Writing',
  Editing = 'Editing',
  Printing = 'Printing',
  Published = 'Published',
  OutOfPrint = 'OutOfPrint',
}

File: csharp/Slimmed/PublishingHouse.Domain/Books/Book.cs
using PublishingHouse.Books.Draft;
using PublishingHouse.Books.Entities;
using PublishingHouse.Books.Factories;
using PublishingHouse.Books.InPrint;
using PublishingHouse.Books.OutOfPrint;
using PublishingHouse.Books.Published;
using PublishingHouse.Books.UnderEditing;
using PublishingHouse.Core.ValueObjects;
using static PublishingHouse.Books.UnderEditing.UnderEditingEvent;
using static PublishingHouse.Books.InPrint.InPrintEvent;
using static PublishingHouse.Books.OutOfPrint.OutOfPrintEvent;

namespace PublishingHouse.Books;

public abstract record Book
{
    public enum State { Writing, Editing, Printing, Published, OutOfPrint }

    public static Book Evolve(Book book, BookEvent @event) =>
        @event switch
        {
            DraftEvent draftEvent => book is InitialBook
                ? BookDraft.Evolve(BookDraft.Default, draftEvent)
                : book,

            MovedToEditing movedToEditing => book is BookDraft
                ? BookUnderEditing.Evolve(BookUnderEditing.Initial, movedToEditing)
                : book,

            UnderEditingEvent underEditingEvent => book is BookUnderEditing underEditing
                ? BookUnderEditing.Evolve(underEditing, underEditingEvent)
                : book,

            MovedToPrinting movedToPrinting => book is BookUnderEditing
                ? BookInPrint.Evolve(BookInPrint.Initial, movedToPrinting)
                : book,

            PublishedEvent.Published published => book is BookInPrint
                ? PublishedBook.Evolve(PublishedBook.Initial, published)
                : book,

            MovedToOutOfPrint movedToOutOfPrint => book is PublishedBook
                ? BookOutOfPrint.Evolve(BookOutOfPrint.Initial, movedToOutOfPrint)
                : book,

            _ => book
        };

    public class Factory: IBookFactory
    {
        public Book Create(
            State state,
            Title title,
            Author author,
            Genre? genre,
            ISBN? isbn,
            CommitteeApproval? committeeApproval,
            List<Reviewer> reviewers,
            List<Chapter> chapters,
            List<Translation> translations,
            List<Format> formats
        ) =>
            state switch
            {
                State.Writing =>
                    new BookDraft(genre, chapters.Select(ch => ch.Title).ToList()),
                State.Editing =>
                    new BookUnderEditing(
                        genre,
                        isbn != null,
                        committeeApproval != null,
                        reviewers.Select(r => r.Id).ToList(),
                        translations.Select(t => t.Language.Id).ToList(),
                        formats.Select(f => (f.FormatType, f.TotalCopies)).ToList()
                    ),
                State.Printing =>
                    new BookInPrint(new PositiveInt(formats.Sum(f => f.TotalCopies.Value))),
                State.Published =>
                    new PublishedBook(
                        new PositiveInt(formats.Sum(f => f.TotalCopies.Value)),
                        new NonNegativeNumber(formats.Sum(f => f.SoldCopies.Value))
                    ),
                State.OutOfPrint =>
                    new BookOutOfPrint(),
                _ => throw new InvalidOperationException()
            };
    }
}
